### PR TITLE
[SPARK-28293][SQL] Implement Spark's own GetTableTypesOperation

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

The table type is from Hive now. This will have some issues. For example, we don't support `index_table`, different Hive supports different table types:
Build with Hive 1.2.1:
image
Build with Hive 2.3.5:
image

This pr implement Spark's own `GetTableTypesOperation`.

## How was this patch tested?

unit tests and manual tests:
image

### CODE REVIEW

**Comment by wangyum** [comment]
@juliuszsompolski What do you think of this change?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        }
+      }
+      setState(OperationState.FINISHED)
+    } catch {
+      case e: HiveSQLException =>
+        setState(OperationState.ERROR)
+        listener.onStatementError(statementId, e.getMessage, SparkUtils.exceptionString(e))
+        throw e
+    }
+    listener.onStatementFinish(statementId)
```
Comment (juliuszsompolski) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala]: add onStatementClosed once https://github.com/apache/spark/pull/25062 is merged.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      statementId,
+      parentSession.getSessionHandle.getSessionId.toString,
+      logMsg,
+      statementId,
+      parentSession.getUsername)
+
+    try {
+      CatalogTableType.tableTypes.foreach { tableType =>
+        if (tableType == EXTERNAL || tableType == EXTERNAL) {
+          rowSet.addRow(Array[AnyRef]("TABLE"))
```
Comment (juliuszsompolski) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala]: I think you meant (tableType == EXTERNAL || tableType == MANAGED), but then you want to add "TABLE" to the results only once.
I think you want
`tableTypes.foreach { tableType => tableTypeString(tableType) }.toSet.foreach { type => rowset.addRow(Array[AnyRef](type) }`

tableTypeString can be shared with SparkGetTablesOperation.
To share some such functions between the operatoins, how about a mixin utils trait
```
trait SparkOperationUtils { this: Operation =>
  def tableTypeString(tableType: CatalogTableType) = ...
}
```
e.g. at the bottom of `SparkSQLOperationManager.scala`?

Comment (wangyum) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala]: +1

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/server/SparkSQLOperationManager.scala [hunk 3]
```diff
@@ -108,6 +108,17 @@ private[thriftserver] class SparkSQLOperationManager()
     operation
   }
 
+  override def newGetTableTypesOperation(
+      parentSession: HiveSession): GetTableTypesOperation = synchronized {
+    val sqlContext = sessionToContexts.get(parentSession.getSessionHandle)
+    require(sqlContext != null, s"Session handle: ${parentSession.getSessionHandle} has not been" +
+      s" initialized or had already closed.")
```
Comment (viirya) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/server/SparkSQLOperationManager.scala]: nit: no string interpolation here.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala (hunk 1 above)
Comment (wangyum) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala]: Sure

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/server/SparkSQLOperationManager.scala (hunk 3 above)
Comment (wangyum) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/server/SparkSQLOperationManager.scala]: OK. Thank you.

**Review by juliuszsompolski — APPROVED** [review]
LGTM pending merge with #25062 (whichever comes first).
cc @gatorsmile

**Comment by wangyum** [comment]
@juliuszsompolski Do we need to implement `GetFunctionsOperation` and `GetTypeInfoOperation`?
image

**Comment by juliuszsompolski** [comment]
Hi @wangyum,
I would have to research a bit, but
- SparkGetFunctionsOperations seems like an useful addition, modeled after SHOW FUNCTIONS / DESCRIBE FUNCTION
- SparkGetTypeInfoOperation... I'm not sure. I think Spark type should be compatible with Hive types, otherwise we would be in trouble because of using Hive result set serialization?

As for others:
- I think we don't need GetCatalogs, am I right to assume it will just return null right now?
- I think we could override GetPrimaryKeys and GetCrossReference operations to just return empty, as even if Spark is connected to an external catalog that contains such reference info, Spark does not utilize these, so better not return them to the tool that may assume that Spark respects these constraints.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ * @param parentSession a HiveSession from SessionManager
+ */
+private[hive] class SparkGetTableTypesOperation(
+    sqlContext: SQLContext,
+    parentSession: HiveSession)
+  extends GetTableTypesOperation(parentSession) with SparkMetadataOperationUtils with Logging {
+
+  override def runInternal(): Unit = {
+    val statementId = UUID.randomUUID().toString
+    val logMsg = s"Listing table types"
```
Comment (HyukjinKwon) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala]: nit .. `s`.

Comment (wangyum) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetTableTypesOperation.scala]: OK. Thank you.

**Comment by gatorsmile** [comment]
The PR https://github.com/apache/spark/pull/25062 has been merged. Could you update this PR?

**Comment by juliuszsompolski** [comment]
LGTM. Thanks!
(sorry for the delay, missed that it's been updated)

**Comment by juliuszsompolski** [comment]
@gatorsmile WDYT about https://github.com/apache/spark/pull/25073#issuecomment-509273235 ?

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
SparkGetFunctionsOperations is useful, but I am also not sure about SparkGetTypeInfoOperation. Do it only when somebody is asking for it?

**Review by gatorsmile — APPROVED** [review]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**